common/scala/src/main/scala/whisk/core/containerpool/logging/SplunkLogStore.scala

```diff
@@ -0,0 +1,158 @@
+    extends LogDriverLogStore(actorSystem) {
+  implicit val as = actorSystem
+  implicit val ec = as.dispatcher
+  implicit val materializer = ActorMaterializer()
+
+  val splunkHost = config.getString("whisk.logstore.splunk.host")
+  val splunkPort = config.getInt("whisk.logstore.splunk.port")
+  private val splunkApi = "/services/search/jobs" //see http://docs.splunk.com/Documentation/Splunk/6.6.3/RESTREF/RESTsearch#search.2Fjobs
+  private val splunkUser = config.getString("whisk.logstore.splunk.user")
+  private val splunkPass = config.getString("whisk.logstore.splunk.password")
+  private val splunkIndex = config.getString("whisk.logstore.splunk.index")
+  private val logMessageFieldName = config.getString("whisk.logstore.splunk.log-message-field")
+  private val activationIdFieldName = config.getString("whisk.logstore.splunk.activation-id-field")
+  private val disableSNI = config.getBoolean("whisk.logstore.splunk.disableSNI")
+
+  val log = actorSystem.log
+  val maxPendingRequests = 500
+
+  val defaultHttpFlow = Http().cachedHostConnectionPoolHttps[Promise[HttpResponse]](
+    host = splunkHost,
+    port = splunkPort,
+    connectionContext =
+      if (disableSNI)
+        Http().createClientHttpsContext(AkkaSSLConfig().mapSettings(s => s.withLoose(s.loose.withDisableSNI(true))))
+      else Http().defaultClientHttpsContext)
+
+  override def fetchLogs(activation: WhiskActivation): Future[ActivationLogs] = {
+
+    //example curl request:
+    //    curl -u  username:password -k https://splunkhost:port/services/search/jobs -d exec_mode=oneshot -d output_mode=json -d "search=search index=\"someindex\" | spath=activation_id | search activation_id=a930e5ae4ad4455c8f2505d665aad282 |  table log_message" -d "earliest_time=2017-08-29T12:00:00" -d "latest_time=2017-10-29T12:00:00"
+    //example response:
+    //    {"preview":false,"init_offset":0,"messages":[],"fields":[{"name":"log_message"}],"results":[{"log_message":"some log message"}], "highlighted":{}}
+    val search =
+      s"""search index="${splunkIndex}"| spath ${activationIdFieldName} | search ${activationIdFieldName}=${activation.activationId.toString} | table ${logMessageFieldName}"""
+
+    val formatter = DateTimeFormatter.ofPattern("YYYY-MM-dd'T'hh:mm:ss").withZone(ZoneId.of("UTC"))
+    val entity = FormData(
+      Map(
+        "exec_mode" -> "oneshot",
+        "search" -> search,
+        "output_mode" -> "json",
+        "earliest_time" -> formatter.format(activation.start),
+        "latest_time" -> formatter.format(activation.end))).toEntity
+
+    log.debug("sending request")
+    queueRequest(
+      Post(splunkApi)
+        .withEntity(entity)
+        .withHeaders(List(Authorization(BasicHttpCredentials(splunkUser, splunkPass))))).flatMap(response => {
+      log.debug(s"splunk API response ${response}")
+      if (response.status.isSuccess()) {
+        Unmarshal(response.entity)
+          .to[String]
```

Comment:
Is it possible to define the log format statically and Unmarshal straight to that?

For example:

```scala
case class SplunkResponse(results: Seq[SplunkResult])
case class SpunkResult(logMessage: String)
```

You would then be able to use

```scala
Unmarshal(response.entity).to[SplunkResponse].map { response =>
  val messages = response.results.map(_.logMessage)
  ActivationLogs(messages)
}
```

Reply:
I'm not sure, since statically defining the format implies that the log fields (and field names) are the same for everyone. I think this opens a bigger question I have, which is: can we require action developers to leverage structured loggers instead of just printing to stdout/stderr?

Currently this impl assumes "some structure", but does not dictate the field names that are used (they can be specified via config).

I'm open to ideas on this.

Reply:
Hmm, does the fluentd driver not produce the same results for every line written to stdout/stderr? That is: Won't the message being printed to either stream always be captured in the `log_message` field regardless of what the user does?

Upon inspection, fluentd actually [hardcodes "log"](https://github.com/moby/moby/blob/master/daemon/logger/fluentd/fluentd.go#L148-L152) as its output field, maybe Splunk applies the transformation? In any case: How would the user be able to alter that?